### ISSUE TITLE
Strangely time-consuming in solving `sesolve` compared to Qutip

### ISSUE THREAD

**Issue by ZhuangMaMS** [dynamiqs/dynamiqs#785]
When I used Qutip to write the codes to solve `sesolve`, it spends *3 faster than dynamiqs (1.78 seconds vs 4.77 seconds). But when I used dynamiqs to write the same codes to solve `sesolve`, it spends *8 faster than Qutip(4.85 seconds vs 37.4 seconds).
-   use Qutip
```python
import dynamiqs as dq
import jax.numpy as jnp
import numpy as np
import qutip as qt
tf=100
tlist=np.linspace(0,tf,200*int(tf)+1)
Nmax = 5
a1 = qt.tensor(qt.destroy(Nmax), qt.qeye(Nmax), qt.qeye(Nmax))
a1dag = a1.dag()

ac = qt.tensor(qt.qeye(Nmax), qt.destroy(Nmax), qt.qeye(Nmax))
acdag = ac.dag()

a2 = qt.tensor(qt.qeye(Nmax), qt.qeye(Nmax), qt.destroy(Nmax))
a2dag = a2.dag()
# Anharmonicity
alpha1 = -0.200*2*np.pi
alpha2 = -0.200*2*np.pi
alphac = -0.200*2*np.pi

g1c = 0.100*2*np.pi
g2c = 0.100*2*np.pi
g12 = 0.005*2*np.pi

w1 = 6.000*2*np.pi
w2 = 5.000*2*np.pi
wc = 7*2*np.pi 

H_parking = (w1*a1dag*a1 +w2*a2dag*a2+ wc*acdag*ac       
+ 0.5*alpha1*a1dag*a1dag*a1*a1
+ 0.5*alpha2*a2dag*a2dag*a2*a2 
+ 0.5*alphac*acdag*acdag*ac*ac
+ g1c*(a1dag*ac + a1*acdag+a1*ac+a1dag*acdag)
+ g2c*(a2dag*ac + a2*acdag+a2*ac+a2dag*acdag)
+ g12*(a1dag*a2 + a1*a2dag+a1*a2+a1dag*a2dag))

```
```python
%%time
vet_state = qt.tensor(qt.basis(Nmax, 0), qt.basis(Nmax, 0),
qt.basis(Nmax, 0))
tf = 100
tlist = np.linspace(0, tf, 200*int(tf)+1)
Hf = H_parking
output = qt.sesolve(Hf, vet_state, tlist)
fina_state_qt = output.states[-1]
# Wall time: 1.77 s
```
```python
%%time
vet_state=jnp.array(qt.tensor(qt.basis(Nmax, 0), qt.basis(Nmax, 0),
qt.basis(Nmax, 0)).full())
tlist = jnp.linspace(0, tf, 200 * int(tf) + 1)
Hf = jnp.array(H_parking.full())
output = dq.sesolve(Hf, vet_state, tlist)
fina_state_jax = output.states[-1]
# Wall time: 4.77 s
```
But 
-   use dynamiqs 
```python
dq.set_precision('double')  # 'simple' by default

tf=100
tlist = jnp.linspace(0,tf,200*int(tf)+1)

Nmax = 5
a1 = dq.tensor(dq.destroy(Nmax), dq.eye(Nmax), dq.eye(Nmax))
a1dag = dq.dag(a1)
    
ac = dq.tensor(dq.eye(Nmax), dq.destroy(Nmax), dq.eye(Nmax))
acdag = dq.dag(ac)

a2 = dq.tensor(dq.eye(Nmax), dq.eye(Nmax), dq.destroy(Nmax))
a2dag = dq.dag(a2)
    
    
# Anharmonicity
  
alpha1 = -0.200 * 2 * jnp.pi
alpha2 = -0.200 * 2 * jnp.pi
alphac = -0.200 * 2 * jnp.pi

g1c = 0.100 * 2 * jnp.pi
g2c = 0.100 * 2 * jnp.pi
g12 = 0.005 * 2 * jnp.pi

w1 = 6.000 * 2 * jnp.pi
w2 = 5.000 * 2 * jnp.pi
wc = 7.000 * 2 * jnp.pi
    
H_parking = (w1*a1dag@a1 +w2*a2dag@a2+ wc*acdag@ac       
+ 0.5*alpha1*a1dag@a1dag@a1@a1
+ 0.5*alpha2*a2dag@a2dag@a2@a2 
+ 0.5*alphac*acdag@acdag@ac@ac
+ g1c * (a1dag@ac + a1@acdag+a1@ac+a1dag@acdag)
+ g2c * (a2dag@ac + a2@acdag+a2@ac+a2dag@acdag)
+ g12 * (a1dag@a2 + a1@a2dag+a1@a2+a1dag@a2dag))
```

```python
%%time
vet_state = dq.tensor(dq.basis(Nmax, 0), dq.basis(Nmax, 0),
dq.basis(Nmax, 0))
tlist = jnp.linspace(0, tf, 200 * int(tf) + 1)
Hf = H_parking
output = dq.sesolve(Hf, vet_state, tlist)
fina_state_jax = output.states[-1]
# Wall time: 4.85 s
```

```python
%%time
vet_state = dq.tensor(dq.basis(Nmax, 0), dq.basis(Nmax, 0),
dq.basis(Nmax, 0))
tlist = jnp.linspace(0, tf, 200 * int(tf) + 1)
Hf = H_parking
output = dq.sesolve(Hf, vet_state, tlist)
fina_state_jax = output.states[-1]
#Wall time: 37.4 s
```
My environment In jupyter
QuTiP Version:      5.0.3.post1
Numpy Version:      1.26.0
Scipy Version:      1.13.1
Cython Version:     0.29.34
Matplotlib Version: 3.7.1
Python Version:     3.9.16
Number of CPUs:     24
BLAS Info:          Generic
INTEL MKL Ext:      False
Platform Info:      Windows (AMD64)

**Comment by gautierronan** [comment]
Hi @ZhuangMaMS, I can reproduce the first two benchmarks, but not the last two. I am not seeing a x8 slowdown (37.4 s) by using dynamiqs for system initialization, even in double precision. On my laptop it's still the same order of magnitude (~4-5 seconds).

For the first benchmarks, I would say there are two reasons:

 - Dynamiqs is using dense arrays instead of sparse ones, but this will be fixed with v0.3 (coming soon). You can also try it out on the branch `dev-qarray`.
 - Your problem is a closed-system with fast time-dynamics on a rather small Hilbert space (125). This is really not where Dynamiqs shines the most, because we're limited by the JAX compiler whereas QuTiP is written in very optimised Cython+Scipy solvers. You will see much better speedups if you make use of:
   - Open systems
   - Batching
   - Larger Hilbert spaces

If you can provide a shorter and complete MWE for your second benchmark, I would be happy to investigate further.

**Comment by ZhuangMaMS** [comment]
Hi @gautierronan, Thanks for your help! In the second benchmarking, I copied the wrong code. This should be
```python 
%%time
vet_state = qt.Qobj(dq.to_qutip(dq.tensor(dq.basis(Nmax, 0), dq.basis(Nmax, 0),
                    dq.basis(Nmax, 0))), dims=[[5, 5, 5],[1,1,1]])

tf = 100
tlist = np.linspace(0, tf, 200*int(tf)+1)
Hf =  qt.Qobj(dq.to_qutip(H_parking), dims=[[5, 5, 5],[5, 5, 5]])
output = qt.sesolve(Hf, vet_state, tlist)
fina_state_qt = output.states[-1]
#Wall time: 37.4 s
```
I found that the reason for being so slow is that using `dq.to_qutip` will convert the Hamiltonian to dense form, while it is actually diagonal here. So I should change the code to `Hf =  qt.Qobj(dq.to_qutip(H_parking), dims=[[5, 5, 5],[5, 5, 5]]).to("Dia")`or `.to("CSR")`.
